### PR TITLE
Update hpctoolkit to latest version with GPU support

### PR DESCRIPTION
**Pull request by pramodk**
This PR updates hpctoolkit and it's dependencies to upstream Spack. It's required to have latest GPU profiling support that we have been trying on BB5 with the help of HPCToolkit developers.
Only changes specific to BBP spack are in fb2952a08811dacf41d775239b0cf0e4e168278d:

 * Add latest master as a release
     - add patch provided by Mark for HPE-MPI
     - enable gpu support in deployment
     ~- pin dyninst version as well as latest one has bug~

Other two commits copy dependencies and hpctoolkit changes:

* Also update rocm packages from upstream
*  Update hpctoolkit to latest version from upstream
      - also update major dependencies of hpctoolkit to be consistent
      - enable latest version of hpctoolkit with GPU support
* binutils has gold variant enabled now. fix `py-llvmlite` to avoid concretisation conflict

~So in principle, these 3 commits can be rebased instead of squash.~

### CODE REVIEW

File: var/spack/repos/builtin/packages/dyninst/package.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/dyninst/package.py:1]: This is a ginormous update in this file :)

File: var/spack/repos/builtin/packages/hip/package.py [hunk 2]
```diff
@@ -13,29 +14,34 @@ class Hip(CMakePackage):
...
+    version('3.10.0', sha256='0082c402f890391023acdfd546760f41cb276dffc0ffeddc325999fd2331d4e8')
     version('3.9.0', sha256='25ad58691456de7fd9e985629d0ed775ba36a2a0e0b21c086bd96ba2fb0f7ed1')
     version('3.8.0', sha256='6450baffe9606b358a4473d5f3e57477ca67cff5843a84ee644bcf685e75d839')
     version('3.7.0', sha256='757b392c3beb29beea27640832fbad86681dbd585284c19a4c2053891673babd')
     version('3.5.0', sha256='ae8384362986b392288181bcfbe5e3a0ec91af4320c189bd83c844ed384161b3')
 
     depends_on('cmake@3:', type='build')
     depends_on('perl@5.10:', type=('build', 'run'))
-    depends_on('mesa~llvm@18.3:')
+    depends_on('mesa18~llvm@18.3:')
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/hip/package.py:33]: ew. no likey. but I guess this is upstream.

File: var/spack/repos/builtin/packages/hpctoolkit/package.py [hunk 3]
```diff
@@ -139,3 +175,18 @@ def configure_args(self):
...
         return args
+
+    # Remove setenv of ROCM, HIP, etc from the module file.  Loading
+    # the hpctoolkit module is not relevant to building a GPU app and
+    # some variables (HIP_PATH) intefere with building the app.
+    def setup_run_environment(self, env):
+        keeplist = []
+        for elt in env.env_modifications:
+            if not (isinstance(elt, SetEnv)
+                    and (elt.name.find('ROCM') >= 0
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/hpctoolkit/package.py]: `"ROCM" not in elt.name` to be more pythonic?

File: var/spack/repos/builtin/packages/roctracer-dev/package.py [hunk 4]
```diff
@@ -42,7 +46,7 @@ def patch(self):
                     string=True)
         kwargs = {'ignore_absent': False, 'backup': False, 'string': False}
         with working_dir('script'):
-            match = '^#!/usr/bin/python'
+            match = '^#!/usr/bin/python[2]'
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/roctracer-dev/package.py]: is this from upstream? shouldn't `[2] == 2` for regular expressions? Seems very fishy.

**Review by matz-e — COMMENTED** [review]
I'd approve if it weren't for that `python[2]` regexp.

File: var/spack/repos/builtin/packages/roctracer-dev/package.py (hunk 4 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/roctracer-dev/package.py]: yeah...I have just checked out this from upstream. I will take a look at why this is done this way.

File: var/spack/repos/builtin/packages/dyninst/package.py (hunk 1 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/dyninst/package.py:1]: synced with upstream again and you see more updates

File: var/spack/repos/builtin/packages/hip/package.py (hunk 2 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/hip/package.py:33]: yeah, all stuff is coming from upstream

File: var/spack/repos/builtin/packages/hpctoolkit/package.py (hunk 3 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/hpctoolkit/package.py]: we can leave that as it is to avoid conflict with upstream?

**Comment by pramodk** [comment]
retest this please

**Comment by pramodk** [comment]
> I'd approve if it weren't for that python[2] regexp.

this is addressed in the upstream changes (with `[23]`). Rest of the changes are from upstream and I kept them as it is to avoid future conflicts.

**Review by matz-e — DISMISSED** [review]

**Comment by pramodk** [comment]
Tested on BB5 and working fine.

**Comment by pramodk** [comment]
retest this please

**Comment by pramodk** [comment]
retest this please

**Comment by pramodk** [comment]
Retest this please

**Comment by pramodk** [comment]
Retest this please

**Comment by matz-e** [comment]
We currently have `hpctoolkit/2021.03.01`…

**Pull request closed without merging**